### PR TITLE
feat: publish stubs

### PR DESCRIPTION
**Pull request by nathane**
This PR allows users to publish all of Blueprint's stubs to `/stubs/blueprint` directory for customization.

Usage:
`php artisan vendor:publish --tag blueprint-stubs`

It currently does not use Laravel 7's stubs. #164

### CODE REVIEW

File: src/BlueprintServiceProvider.php [hunk 1]
```diff
@@ -20,9 +20,17 @@ class BlueprintServiceProvider extends ServiceProvider implements DeferrableProv
     public function boot()
     {
         if (!defined('STUBS_PATH')) {
-            define('STUBS_PATH', dirname(__DIR__) . '/stubs');
+            $stubPath = base_path('stubs/blueprint');
```
Comment (jasonmccreary) [inline src/BlueprintServiceProvider.php:23]: Is `stubs` the path where Laravel _would_ save to?

Comment (nathane) [reply src/BlueprintServiceProvider.php:23]: that's correct.

**Comment by jasonmccreary** [comment]
I'd like to take at least one more step on this to the ultimate direction of sharing these templates with anything the developer published in Laravel.

Let's update our stub _placeholders_ to match what Laravel uses for the namespace, class, etc. I believe it's more of a blade syntax.

**Comment by nathane** [comment]
I agree. This PR was meant to satisfy the current request.

I was working on a separate PR, changing the current stubs in Blueprint to match Laravel’s customizable stubs. 

I've been using `str_replace` and `preg_replace` to populate the stubs. I’m not sure this approach is adequate, especially after users make changes to the default Laravel stubs. https://github.com/nathane/blueprint/tree/feature/refactoring-stubs

**Comment by jasonmccreary** [comment]
I've seen Taylor mention this, but I'll take a closer look.

Long story short is I think if you modify a stub, it's _understood_ that you need to maintain the placeholders.

**Comment by nathane** [comment]
Not sure if I got my point across, I’m trying to say if the users change something in the stubs, that may affect the regular expression Blueprint uses to generate the correct code.

Blueprint also has a few placeholder (eg. Doc blocks, partials) that are not in Laravel’s customizable stubs. Let me know if you have any suggestions on how to handle those without breaking the current functionality.

**Comment by jasonmccreary** [comment]
I understand.

I think this PR should align with Laravel where we do intersect and then point out where we don't.

From there we can refactor Blueprint to hook into the code in alternative ways. I'm glad to help with that, but if you could get the ball rolling, it'd be a big help.

**Comment by nathane** [comment]
Sure, I’ll finish what I’ve started.

**Comment by jasonmccreary** [comment]
@nathane, this looks good and all tests are passing. Is this ready to merge or did you have more you wanted to add?

While not necessarily part of this PR, my understanding of what is left towards the overall goal would be to fully support existing Laravel 7 stubs the user may have published within their own projects.

**Comment by nathane** [comment]
No, it’s not ready to merge. 

This pr is still publishing Blueprint’s stubs in `stubs/blueprint ` and not utilizing Laravel’s stubs.

I’ve refactored the stubs to match Laravel’s, using unconventional methods to get the tests passing and “get the ball rolling”, you may take over from here.

**Comment by jasonmccreary** [comment]
Fair enough, I'll likely keep this open and attempt to commit to it directly. Do you know if you checked that special _maintainer edits_ box 😅 

Also, @ttomdewit had previously mentioned a package which is nice for editing existing PHP classes. Maybe it's time to adopt that.

**Comment by nathane** [comment]
Yes, “Allow edits by maintainers” is checked.

**Comment by ttomdewit** [comment]
> Fair enough, I'll likely keep this open and attempt to commit to it directly. Do you know if you checked that special _maintainer edits_ box 😅
> 
> Also, @ttomdewit had previously mentioned a package which is nice for editing existing PHP classes. Maybe it's time to adopt that.

For reference that would probably be https://github.com/nette/php-generator

**Comment by jasonmccreary** [comment]
@ttomdewit, do you know if you can load in an existing class with this? Or is it only creating new classes?

**Comment by nathane** [comment]
How about https://github.com/nikic/PHP-Parser ?

**Comment by ttomdewit** [comment]
> @ttomdewit, do you know if you can load in an existing class with this? Or is it only creating new classes?

I think you’re looking for https://github.com/nette/php-generator/blob/master/readme.md#generate-using-reflection

**Comment by ttomdewit** [comment]
> How about https://github.com/nikic/PHP-Parser ?

I think this one is for manipulating existing code. Might be worthwhile if Nette’s generation based on existing code isn’t sufficient.

**Comment by jasonmccreary** [comment]
I created [this repo](https://github.com/jasonmccreary/laravel-stubs) containing the latest Laravel stubs as a reference to help complete this PR.

**Comment by Algeneralo** [comment]
@jasonmccreary is there any updates here?

**Comment by jasonmccreary** [comment]
Nope. Probably going to close it given the age and conflicts. You are welcome to try and restart.

**Pull request closed without merging**